src/enumerable.cr

```diff
@@ -1912,9 +1912,20 @@ module Enumerable(T)
     end
   end
 
-  # :nodoc:
-  def first_internal
-    # overridden in Tuple to disable literal index lookup
-    first
+  # Returns a value with the same type as an element of *x*, even if *x* is not
+  # an `Enumerable`.
+  #
+  # Used by splat expansion inside array literals. For example, this code
+  #
+  # ```
+  # [1, *{2, 3.5}, 4]
+  # ```
+  #
+  # will end up calling `typeof(1, ::Enumerable.element_type({2, 3.5}), 4)`.
+  #
+  # NOTE: there should never be a need to call this method outside the standard library.
+  def self.element_type(x)
+    x.each { |elem| return elem }
+    ::raise ""
```

Comment:
This method is intended for being used in `typeof` and not supposed to be actually called. But still, it can be. And then it should raise a proper exception, instead of an empty message.

Unfortunately it's not possible to move the `typeof` into the method because of nesting calls and yielded values. That would probably be the best solution.

Reply:
Do we want `element_type` as a public method like this?

I realize it's already merged from #10429, thus not necessarily subject to this PR.

Reply:
It cannot be even protected because splat expansions must be able to access the method in any scope, for the same reason as `::String.interpolation` being public.

Reply:
I think it's actually good to have this public because it enables shards to use it if they need it.

Reply:
But what about the empty raise? It's public API and can be called outside of `typeof`.

Reply:
Either `EmptyError` or a generic `BUG:` exception, I guess.